cmd/bcpd/app/app_test.go

```diff
@@ -225,4 +246,86 @@ func TestApp(t *testing.T) {
 	require.NoError(t, err)
 	assert.Equal(t, int32(3), toke.SigFigs)
 	assert.Equal(t, "Frankie", toke.Name)
+
+	// create contract
+	pk = crypto.GenPrivKeyEd25519()
+	pk2 = crypto.GenPrivKeyEd25519()
+	pk3 := crypto.GenPrivKeyEd25519()
+	signers := [][]byte{pk.PublicKey().Address(), pk2.PublicKey().Address(), pk3.PublicKey().Address()}
+
+	cmsg := &multisig.CreateContractMsg{
+		Sigs:                signers,
+		ActivationThreshold: 2,
+		AdminThreshold:      2,
+	}
+	tx = &Tx{
+		Sum: &Tx_CreateContractMsg{cmsg},
+	}
+	sig1, err := sigs.SignTx(pk, tx, chainID, 0)
+	require.NoError(t, err)
+	tx.Signatures = []*sigs.StdSignature{sig1}
+	txBytes, err = tx.Marshal()
+	require.NoError(t, err)
+	require.NotEmpty(t, txBytes)
+
+	// Submit to the chain
+	header = abci.Header{Height: 3}
+	myApp.BeginBlock(abci.RequestBeginBlock{Header: header})
+	// check and deliver must pass
+	chres = myApp.CheckTx(txBytes)
+	require.Equal(t, uint32(0), chres.Code, chres.Log)
+	dres = myApp.DeliverTx(txBytes)
+	require.Equal(t, uint32(0), dres.Code, dres.Log)
+	// make sure commit is proper
+	myApp.EndBlock(abci.RequestEndBlock{})
+	cres = myApp.Commit()
+	block1 = cres.Data
+	assert.NotEmpty(t, block1)
+	assert.Equal(t, chainID, myApp.GetChainID())
+
+	// get a contract
+	cquery := abci.RequestQuery{
+		Path: "/contracts",
+		Data: contractID,
+	}
+	var c multisig.Contract
+	cqres := myApp.Query(cquery)
+	err = app.UnmarshalOneResult(cqres.Value, &c)
+	require.NoError(t, err)
+	assert.Equal(t, signers, c.Sigs)
+	assert.EqualValues(t, 2, c.ActivationThreshold)
+	assert.EqualValues(t, 2, c.AdminThreshold)
+
+	// build and sign a transaction with contract
+	msg = &cash.SendMsg{
+		Src:  contractAddr,
+		Dest: addr2,
+		Amount: &x.Coin{
+			Whole:  2000,
+			Ticker: "ETH",
+		},
+		Memo: "Have a great trip!",
+	}
+	tx = &Tx{
+		Sum:      &Tx_SendMsg{msg},
+		Multisig: contractID,
+	}
+	sig1, err = sigs.SignTx(pk, tx, chainID, 1)
```

Comment:
The following sequence of lines is repeated many times (also in my earlier code).
Please simplify this (in another commit, or another PR based on top).

```
type Signer { pk: PrivateKey, nonce: int64 }
func signAndCommit(t testing.T, tx Tx, signers []Signer, chainId string, blockHeight int64) DeliverTxResult
```

it can contain all the assertions and all the steps to ensure a correct execution (or test failure) and provide DeliverTxResult for any further testing.